src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs

```diff
@@ -483,6 +486,25 @@ private static bool TryGetVersionRange(IVsReferenceItem item, string propertyNam
             return versionRange != null;
         }
 
+        private static IEnumerable<VersionRange> GetVersionRangeList(IVsReferenceItem item)
+        {
+            char[] splitChars = new[] { ';' };
+            string versionString = GetPropertyValueOrNull(item, "Version");
+
+            if (versionString != null)
+            {
+                var versions = versionString.Split(splitChars, StringSplitOptions.RemoveEmptyEntries);
+                foreach (var version in versions)
+                {
+                    yield return VersionRange.Parse(version);
+                }
+            }
+            else
+            {
+                yield return VersionRange.All;
```

Comment:
Are allocations on this code path a concern? I guess no, because project nominations don't happen so often, and this isn't on the hot path for restore, but I thought it's worth checking.  Methods that use `yield` cause the method to be a state machine, and therefore allocates on the heap, similar to async methods. If we need to hyper-optimize, then changing this method to something like `private static void AddVersionRangeList(IVsReferenceItem item, List<VersionRange> versionList)` will be lower allocations, but perhaps not as nice to read (who doesn't love pure fuctions).

On the other hand, the nomination code uses so much LINQ there are tons of unnecessary allocations (although short lived)

Reply:
PackageDownload are infrequent as well. 

Honestly, it could be a `fun` challenge to write all these to make them performant and as similar as possible on the commandline and VS :D